contracts/src/v0.8/dev/ocr2dr/OCR2DRRegistry.sol

```diff
@@ -0,0 +1,742 @@
+    }
+    // No lower bound on the requested gas limit. A user could request 0
+    // and they would simply be billed for the gas and computation.
+    if (billing.gasLimit > s_config.maxGasLimit) {
+      revert GasLimitTooBig(billing.gasLimit, s_config.maxGasLimit);
+    }
+
+    // Check that subscription can afford the estimated cost
+    uint96 oracleFee = OCR2DROracleInterface(msg.sender).getRequiredFee(data, billing);
+    uint256 estimatedCost = estimateCost(data, billing, oracleFee);
+    if (s_subscriptions[billing.subscriptionId].balance < estimatedCost) {
+      revert InsufficientBalance();
+    }
+
+    uint64 nonce = currentNonce + 1;
+    bytes32 requestId = computeRequestId(msg.sender, billing.client, billing.subscriptionId, nonce);
+
+    Commitment memory commitment = Commitment(
+      billing.subscriptionId,
+      billing.client,
+      billing.gasLimit,
+      msg.sender,
+      oracleFee,
+      getRequiredFee(data, billing)
+    );
+    s_requestCommitments[requestId] = commitment;
+
+    emit BillingStart(requestId, commitment);
+    s_consumers[billing.client][billing.subscriptionId] = nonce;
+    return requestId;
+  }
+
+  function computeRequestId(
+    address don,
+    address client,
+    uint64 subscriptionId,
+    uint64 nonce
+  ) private pure returns (bytes32) {
+    return keccak256(abi.encode(don, client, subscriptionId, nonce));
+  }
+
+  /**
+   * @dev calls target address with exactly gasAmount gas and data as calldata
+   * or reverts if at least gasAmount gas is not available.
+   */
+  function callWithExactGas(
+    uint256 gasAmount,
+    address target,
+    bytes memory data
+  ) private returns (bool success) {
+    // solhint-disable-next-line no-inline-assembly
+    assembly {
+      let g := gas()
+      // GAS_FOR_CALL_EXACT_CHECK = 5000
+      // Compute g -= GAS_FOR_CALL_EXACT_CHECK and check for underflow
+      // The gas actually passed to the callee is min(gasAmount, 63//64*gas available).
+      // We want to ensure that we revert if gasAmount >  63//64*gas available
+      // as we do not want to provide them with less, however that check itself costs
+      // gas.  GAS_FOR_CALL_EXACT_CHECK ensures we have at least enough gas to be able
+      // to revert if gasAmount >  63//64*gas available.
+      if lt(g, 5000) {
+        revert(0, 0)
+      }
+      g := sub(g, 5000)
+      // if g - g//64 <= gasAmount, revert
+      // (we subtract g//64 because of EIP-150)
+      if iszero(gt(sub(g, div(g, 64)), gasAmount)) {
+        revert(0, 0)
+      }
+      // solidity calls check that a contract actually exists at the destination, so we do the same
+      if iszero(extcodesize(target)) {
+        revert(0, 0)
+      }
+      // call and return whether we succeeded. ignore return data
+      // call(gas,addr,value,argsOffset,argsLength,retOffset,retLength)
+      success := call(gasAmount, target, 0, add(data, 0x20), mload(data), 0, 0)
```

Comment:
If I understand correctly, this doesn't let out any reverts that happen inside, which messes up with our try-catch logic... Let's leave a TODO to maybe at least think about it. cc @pinebit